sound/soc/sof/topology.c

```diff
@@ -1275,11 +1383,67 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 	swidget->id = w->id;
 	swidget->pipeline_id = index;
 	swidget->private = NULL;
+	swidget->num_sink_pins = 1;
```

Comment:
how about host copier and dai copier ?  do they have both source & sink pin ?

Reply:
I have explained this in the design document in detail. there will be no ASoC route defined for the widget that has zero source/sink pin, and a zero value is dangerous for ida_alloc_max. For for the zero pin edge case, we still define this value as one, but the value is not used.

Reply:
it's rather convoluted @aiChaoSONG. There's two controversial points above
a) first there's a misleading value num_sink_pin=1 that's set but not used. That's a recipe for future confusion
b) when the num_sink_pin/source is 1 we still use an IDA to allocate a unique value. That's overkill and over-engineered. Conventional arithmetic tells us with a number of pins equal to one there is already a unique value for the pin_id....

Reply:
@plbossart I was trying to be smart, so that we can simplify topology, yeah, I agree this does bring confusion.
the two controversial points are removed now, document is also updated.
a) num_sink/source_pin = 0 is allowed.
b) if num_sink/source_pin = 1, queue id 0 returned directly, IDA is not used anymore.